awsshell/selectmenu.py

```diff
@@ -0,0 +1,264 @@
+                content=BufferControl(
+                    buffer_name='INFO',
+                    # TODO discuss pygments import voodoo
+                    lexer=PygmentsLexer(find_lexer_class('JSON'))
+                )
+            ),
+        )
+    ])
+
+
+class SelectMenuApplication(Application):
+    """Wrap Application, providing the correct layout, keybindings, etc."""
+
+    def __init__(self, message, options, *args, **kwargs):
+        self.menu_control = SelectMenuControl(options)
+
+        self.kb_manager = KeyBindingManager(
+            enable_system_bindings=True,
+            enable_abort_and_exit_bindings=True
+        )
+        menu_control = self.menu_control
+        options_meta = kwargs.pop('options_meta', None)
+
+        # Return the currently selected option
+        def return_selection(cli, buf):
+            cli.set_return_value(menu_control.get_selection())
+
+        buffers = {}
+
+        def_buf = Buffer(
+            initial_document=Document(u''),
+            accept_action=AcceptAction(return_selection)
+        )
+
+        buffers[DEFAULT_BUFFER] = def_buf
+
+        show_meta = options_meta is not None
+        # Optionally show meta information if present
+        if show_meta:
+            info_buf = Buffer(is_multiline=True)
+            buffers['INFO'] = info_buf
+
+            def selection_changed(cli):
+                info = options_meta[buffers[DEFAULT_BUFFER].text]
+                formatted_info = json.dumps(info, indent=4, sort_keys=True,
+                                            ensure_ascii=False)
+                buffers['INFO'].text = formatted_info
+            def_buf.on_text_changed += selection_changed
+
+        # Apply the correct buffers, key bindings, and layout before super call
+        kwargs['buffers'] = buffers
+        kwargs['key_bindings_registry'] = self.kb_manager.registry
+        kwargs['layout'] = create_select_menu_layout(
+            message,
+            menu_control,
+            show_meta=show_meta
+        )
+        self._bind_keys(self.kb_manager.registry, self.menu_control)
+        super(SelectMenuApplication, self).__init__(*args, **kwargs)
+
+    def _bind_keys(self, registry, menu_control):
+        handle = registry.add_binding
+
+        @handle(Keys.F10)
+        def handle_f10(event):
+            event.cli.set_exit()
+
+        @handle(Keys.Up)
+        @handle(Keys.BackTab)
+        def handle_up(event):
+            menu_control.select_up(event=event)
+
+        @handle(Keys.Tab)
+        @handle(Keys.Down)
+        def handle_down(event):
+            menu_control.select_down(event=event)
+
+        @handle(Keys.ControlJ)
+        def accept(event):
+            buff = event.current_buffer
+            selection = menu_control.get_selection()
+            if selection is not None:
+                buff.accept_action.validate_and_handle(event.cli, buff)
+
+        @handle(Keys.Any)
+        @handle(Keys.Backspace)
```

Comment:
Yes, while it does cover `Keys.Backspace` it will not overwrite the previous binding that were setup in the `KeyBindingManager` initialization, so I must explicitly bind `Keys.Backspace` again to ensure that the previous bindings do not get called.